### PR TITLE
TP 2.0

### PR DESCRIPTION
**Pull request by dlemures**
* [x] support incremental annotation processing. All processors are now incremental and isolating.
* [x] remove all registries, TP 2.x doesn't use registries any more. They were actually slower than not using them after
we reworked on registries to enable obfuscation. Reflection is now used to load the factories and member injectors,
this is normal (and all other annotation processor based libs do that too) and there is no impact on performance.
* [x] support multiple rounds of annotation processing.
* [x] generated classes now use `__` instead of `$$` to make TP more compliant with some tools.
* [x] removed the generate ContextSingleton annotation
* [x] smoothie now binds the ClipClipboardManager service. Thx to Cody Henthrone for this. Sorry we were late merging this !
* [x] smoothie now contains non support and non android x bindings. We have introduced 2 new artifacts:
smoothie-support and smoothie-androidx that respectively support the old legacy support library and the new android X libs).
* [x] toothpick-testing now only contains core testing classes. New artifacts have been introduced to support junit 4 nd junit 5 (resp. : toothpick-testing-junit4,  toothpick-testing-junit5)  
* [x] Update CI & make it green
* [x] Wiki update
* [x] version number update
* [x] test in MGA (snapshot + proguard + artifacts: junit-4, androidx, template new modules)
* [x] go through PR list and issue list
* [x] deal with the PR / feature request  #318 
* [x] README update
* [x] add a consumer proguard file: https://github.com/JakeWharton/butterknife/search?q=proguard-rules.txt&unscoped_q=proguard-rules.txt
* [x] update other libs 
* [ ] ~make the sample work with incremental compilation (there is a bug in my fork of incap that I can repro with smoothie-sample !). This won't be fixed now.~
* [x] make the sample work with junit 5
* [x] make the sample work with android x
* [x] use composite builds for samples
* [x] why do we need to use default configuration ? Is compositve build solving that ?
* [ ] declare the project as incremental on gradle list : https://github.com/gradle/gradle/issues/5277

### CODE REVIEW

File: toothpick-runtime/build.gradle [hunk 1]
```diff
@@ -10,22 +8,23 @@ configurations {
...
   api deps.inject
 
-  testImplementation deps.junit
-  testImplementation project(':toothpick-generated-core')
+  testImplementation deps.junit4
   testImplementation deps.easymock
   testImplementation deps.powermock
   testImplementation deps.hamcrest
-  testImplementation files(Jvm.current().getToolsJar())
+  testImplementation 'org.powermock:powermock-module-junit4-rule-agent:1.6.4'
```
Comment (dlemures) [inline toothpick-runtime/build.gradle]: this should go inside `deps`

File: toothpick-runtime/build.gradle [hunk 2]
```diff
@@ -10,22 +8,23 @@ configurations {
...
   testAnnotationProcessor project(':toothpick-compiler')
 }
 
+test {
+  jvmArgs '-noverify'
+}
```
Comment (dlemures) [inline toothpick-runtime/build.gradle]: should we keep this?

File: toothpick-sample/build.gradle [hunk 3]
```diff
@@ -1,32 +1,21 @@
...
+  testImplementation deps.hamcrest
+  testAnnotationProcessor deps.tp_compiler
 }
-
-afterEvaluate {
-  if (project.tasks.findByName('check')) {
-    jacocoTestReport {
-      reports {
-        xml.enabled = false
-        html.enabled = false
```
Comment (dlemures) [inline toothpick-sample/build.gradle]: dont need this anymore?

File: toothpick-runtime/build.gradle (hunk 1 above)
Comment (dlemures) [reply toothpick-runtime/build.gradle]: done

File: smoothie-androidx/gradle.properties [hunk 4]
```diff
@@ -0,0 +1,4 @@
+POM_ARTIFACT_ID=smoothie-androidx
+POM_NAME=smoothie android X
+POM_DESCRIPTION='Toothpick package providing Android providers for the android X library'
+POM_PACKAGING='aar'
```
Comment (stephanenicolas) [inline smoothie-androidx/gradle.properties]: .

File: smoothie-androidx/src/test/java/toothpick/smoothie/module/SmoothieSupportAndroidXModuleTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    LayoutInflater layoutInflater = activityScope.getInstance(LayoutInflater.class);
+
+    //THEN
+    assertThat(injectedActivity, instanceOf(FragmentActivity.class));
+    assertThat((FragmentActivity) injectedActivity, sameInstance(activity));
+    assertThat(fragmentManager, notNullValue());
+    assertThat(loaderManager, notNullValue());
+    assertThat(layoutInflater, notNullValue());
+  }
+}
```
Comment (stephanenicolas) [inline smoothie-androidx/src/test/java/toothpick/smoothie/module/SmoothieSupportAndroidXModuleTest.java]: .

File: smoothie-sample/src/main/java/com/example/smoothie/SimpleApp.java [hunk 6]
```diff
@@ -4,23 +4,16 @@
 import toothpick.configuration.Configuration;
 import toothpick.Scope;
 import toothpick.Toothpick;
-import toothpick.registries.FactoryRegistryLocator;
-import toothpick.registries.MemberInjectorRegistryLocator;
+import toothpick.locators.FactoryLocator;
+import toothpick.locators.MemberInjectorLocator;
```
Comment (stephanenicolas) [inline smoothie-sample/src/main/java/com/example/smoothie/SimpleApp.java]: not needed

File: smoothie-support/gradle.properties [hunk 7]
```diff
@@ -0,0 +1,4 @@
+POM_ARTIFACT_ID=smoothie-support
+POM_NAME=smoothie support
+POM_DESCRIPTION='Toothpick package providing Android providers for the support library'
+POM_PACKAGING='aar'
```
Comment (stephanenicolas) [inline smoothie-support/gradle.properties]: .

File: toothpick-compiler/src/main/resources/META-INF/gradle/incremental.annotation.processors [hunk 8]
```diff
@@ -0,0 +1,2 @@
+toothpick.compiler.factory.FactoryProcessor,isolating
+toothpick.compiler.memberinjector.MemberInjectorProcessor,isolating
```
Comment (stephanenicolas) [inline toothpick-compiler/src/main/resources/META-INF/gradle/incremental.annotation.processors]: .

File: toothpick-runtime/build.gradle (hunk 2 above)
Comment (stephanenicolas) [reply toothpick-runtime/build.gradle]: yes, powermock breaks the java 7 byte code frames otherwise.

File: toothpick-runtime/build.gradle [hunk 9]
```diff
@@ -10,22 +8,23 @@ configurations {
...
+
 compileTestJava {
-  options.annotationProcessorPath = configurations.testAnnotationProcessor
-  options.compilerArgs = ['-Atoothpick_registry_package_name=toothpick.test',
-                          '-Atoothpick_annotations=toothpick.data.CustomScope',]
+  options.compilerArgs = ['-Atoothpick_annotations=toothpick.data.CustomScope',]
 }
 
 apply from: rootProject.file('gradle/gradle-mvn-push.gradle')
```
Comment (stephanenicolas) [inline toothpick-runtime/build.gradle]: .

File: toothpick-runtime/src/test/java/toothpick/locators/MemberInjectorLocatorTest.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  @Test
+  public void testGetMemberInjector_shouldNotReturnNull_whenTheMemberInjectorIsNotGenerated() {
+    //GIVEN
+    //WHEN
+    MemberInjector<Qurtz> memberInjector = MemberInjectorLocator.getMemberInjector(Qurtz.class);
+
+    //THEN
+    assertThat(memberInjector, nullValue());
+  }
+}
```
Comment (stephanenicolas) [inline toothpick-runtime/src/test/java/toothpick/locators/MemberInjectorLocatorTest.java]: .

File: toothpick-runtime/src/test/java/toothpick/locators/FactoryLocatorTest.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  @Test(expected = NoFactoryFoundException.class)
+  public void testGetFactory_shouldThrowAnException_whenTheFActoryIsNotGenerated() {
+    //GIVEN
+    //WHEN
+    FactoryLocator.getFactory(Qurtz.class);
+
+    //THEN
+    fail("Should throw an exception");
+  }
+}
```
Comment (stephanenicolas) [inline toothpick-runtime/src/test/java/toothpick/locators/FactoryLocatorTest.java]: .

File: toothpick-sample/build.gradle (hunk 3 above)
Comment (stephanenicolas) [reply toothpick-sample/build.gradle]: no

File: toothpick-sample/src/test/java/toothpick/sample/SimpleEntryPointTest.java [hunk 12]
```diff
@@ -54,10 +39,4 @@ public void testMultiply() throws Exception {
     assertThat(result, is(48));
     verify(mockComputer);
   }
-
-  private class TestModule extends Module {
-    TestModule() {
-      bind(Computer.class).toInstance(mockComputer);
-    }
-  }
 }
```
Comment (stephanenicolas) [inline toothpick-sample/src/test/java/toothpick/sample/SimpleEntryPointTest.java]: .

File: toothpick-testing-junit4/gradle.properties [hunk 13]
```diff
@@ -0,0 +1,4 @@
+POM_ARTIFACT_ID=toothpick-testing-junit4
+POM_NAME=Toothpick Testing Junit 4
+POM_DESCRIPTION='Junit 4 Testing support for Toothpick'
+POM_PACKAGING='jar'
```
Comment (stephanenicolas) [inline toothpick-testing-junit4/gradle.properties]: .

File: toothpick-testing-junit5/gradle.properties [hunk 14]
```diff
@@ -0,0 +1,4 @@
+POM_ARTIFACT_ID=toothpick-testing-junit5
+POM_NAME=Toothpick Testing Junit 5
+POM_DESCRIPTION='Junit 5 Testing support for Toothpick'
+POM_PACKAGING='jar'
```
Comment (stephanenicolas) [inline toothpick-testing-junit5/gradle.properties]: .

**Comment by dlemures** [comment]
You are the one that rocks little penguin 🐥

**Pull request merged**